### ISSUE TITLE
SSHClientConnectionOptions is not reusable across asyncssh.connect() calls

### ISSUE THREAD

**Issue by Snawoot** [ronf/asyncssh#222]
Public key auth in fails if `asyncssh.SSHClientConnectionOptions` is reused another client connection. In my case `client_keys` was like `['private_key_filename']`.

Package versions:
```
[user@dt1 rsp]$ venv/bin/pip freeze 
asn1crypto==0.24.0
asyncssh==1.17.0
cffi==1.12.3
cryptography==2.7
pycparser==2.19
six==1.12.0
```

Python version:
```
[user@dt1 rsp]$ venv/bin/python --version
Python 3.7.3
```

OS:
```
[user@dt1 rsp]$ cat /etc/os-release 
NAME=Fedora
VERSION="30 (Thirty)"
ID=fedora
VERSION_ID=30
VERSION_CODENAME=""
PLATFORM_ID="platform:f30"
PRETTY_NAME="Fedora 30 (Thirty)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:30"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f30/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=30
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=30
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
```

Debug log output: https://gist.github.com/Snawoot/35e8d25e10494e018a2bcafe8fe261f2
Probably only logs from `asyncssh` facility is revelant, but rest may give some context.

Some additional info:
Those connections are established concurrently from different coroutines.

**Comment by Snawoot** [comment]
@ronf

Hello,

Adding test case: https://gist.github.com/Snawoot/517f6e3f9b8327ac6a0ad8b4ded913b6#file-test_asyncssh-py-L50-L62

Only first connection succeeds if they share same instance of SSHClientConnectionOptions.

**Comment by ronf** [comment]
Thanks very much! With your example, I was able to reproduce the problem here.

I suspected that the issue here might be related to not making a copy of the client_keys list in the options object and having that list modified during authentication so that the keys are no longer present there when the second authentication occurs. However, in the previous tests I ran, each connect() call was getting its own unique copy of the list.

After a bit of experimenting, I see I was right about this being the problem. In some cases, the list of keys ends up being copied before it is consumed and everything works, but there are a few cases where this doesn't happen and you ran into one of them. The fix is pretty simple. The following change should do it:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index 7b84cf8..31433ad 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -2387,10 +2387,10 @@ class SSHClientConnection(SSHConnection):
         self._password = options.password
 
         self._client_host_keysign = options.client_host_keysign
-        self._client_host_keys = options.client_host_keys
+        self._client_host_keys = list(options.client_host_keys)
         self._client_host = options.client_host
         self._client_username = options.client_username
-        self._client_keys = options.client_keys
+        self._client_keys = list(options.client_keys)
 
         if options.agent_path is not None:
             self._agent = SSHAgentClient(self._loop, options.agent_path)
```

I'll get this checked in to the "develop" branch shortly, along with a new unit test which explicitly tests this, and it will be part of the next release. Thanks for the report, and for your help reproducing it!

**Comment by ronf** [comment]
After a bit more testing, it turns out the fix should be:

```patch
diff --git a/asyncssh/connection.py b/asyncssh/connection.py
index c42d787..758f528 100644
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -2389,10 +2389,12 @@ class SSHClientConnection(SSHConnection):
         self._password = options.password
 
         self._client_host_keysign = options.client_host_keysign
-        self._client_host_keys = options.client_host_keys
+        self._client_host_keys = None if options.client_host_keys is None else \
+                                 list(options.client_host_keys)
         self._client_host = options.client_host
         self._client_username = options.client_username
-        self._client_keys = options.client_keys
+        self._client_keys = None if options.client_keys is None else \
+                            list(options.client_keys)
 
         if options.agent_path is not None:
             self._agent = SSHAgentClient(self._loop, options.agent_path)
```

I also had to update some of the unit tests. I'll have a fix checked into the "develop" branch momentarily.

**Comment by ronf** [comment]
This fix is now available in the "develop" branch as commit b818732.

**Comment by Snawoot** [comment]
Thank you! Also thanks for this project!

**Comment by ronf** [comment]
My pleasure - glad to hear you're finding it useful!

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 1.17.1.
